src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -79,7 +89,212 @@ struct C_EncryptedObjectReadRequest : public Context {
+    ceph::bufferlist* aligned_data;
+    ceph::bufferlist* cmp_data;
+    uint64_t* mismatch_offset;
+    io::ReadExtents extents;
+    Context* onfinish;
+    uint64_t version;
+    C_UnalignedObjectReadRequest<I>* read_req;
+    std::function<void(std::optional<uint64_t>)> write_req_sender;
+
+    C_UnalignedObjectWriteRequest(
+            I* image_ctx, CryptoInterface* crypto, uint64_t object_no,
+            uint64_t object_off, ceph::bufferlist* data,
+            ceph::bufferlist* cmp_data, uint64_t* mismatch_offset,
+            IOContext io_context, int op_flags, int write_flags,
+            std::optional<uint64_t> assert_version,
+            const ZTracer::Trace &parent_trace, int* object_dispatch_flags,
+            uint64_t* journal_tid,
+            Context* on_dispatched) : cct(image_ctx->cct),
+                                      object_off(object_off),
+                                      aligned_data(data),
+                                      cmp_data(cmp_data),
+                                      mismatch_offset(mismatch_offset),
+                                      onfinish(on_dispatched) {
+      auto aligns = crypto->get_pre_and_post_align(
+              object_off, data->length());
+      auto aligned_object_off = object_off;
+      if (aligns.first != 0) {
+        aligned_object_off -= aligns.first;
+        extents.emplace_back(aligned_object_off, aligns.first);
+      }
+      if (aligns.second != 0) {
+        extents.emplace_back(object_off + data->length(),
+                             aligns.second);
+      }
+
+      if (cmp_data != nullptr) {
+        extents.emplace_back(object_off, cmp_data->length());
+      }
+
+      read_req = new C_UnalignedObjectReadRequest<I>(
+              image_ctx, crypto, object_no, &extents, io_context,
+              0, 0, parent_trace, &version, 0, this);
+
+      write_req_sender = [image_ctx, object_no, aligned_object_off, io_context,
+                          op_flags, write_flags, assert_version, journal_tid,
+                          &parent_trace,
+                          this] (std::optional<uint64_t> version) mutable {
+        bool exclusive = write_flags & io::OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE;
+        if (!version.has_value()) {
+          write_flags |= io::OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE;
+          version = assert_version;
+        } else if (exclusive) {
+          onfinish->complete(-EEXIST);
+          return;
+        } else if (assert_version.has_value() &&
+                   version.value() != assert_version.value()) {
+          if (assert_version.value() < version.value()) {
+            onfinish->complete(-ERANGE);
+          } else {
+            onfinish->complete(-EOVERFLOW);
+          }
+          return;
+        }
+
+        auto ctx = new LambdaContext(
+                [this, exclusive, assert_version](int r) {
+                  if (r == -ERANGE && !assert_version.has_value()) {
+                    r = -ERESTART;
```

Comment:
Who is handling `-ERESTART`?

Reply:
If the assert_version or create flag is caused by the crypto layer, and not by the end-user, then I translate the error to ERESTART, since the error was caused by a race in the read-modify-write logic, which can be retried.
I took inspiration from the copyup code, which also returns ERESTART for writes that didn't make it in-time to be included by the copyup.
I assumed that if a write can return ERESTART due to a copyup race, then its also fine to do the same for crypto races.
Checking again now, I see that ObjectRequest is actually restarting such writes for copyup, so I need to handle my restarts as well.

I think I will add `if (r == -ERESTART)` to `C_UnalignedObjectWriteRequest::finish` that will send a fresh copy of itself.

Reply:
Actually, instead of returning -ERESTART, I will just send a fresh copy of the request.